torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -601,9 +607,16 @@ bool ProcessGroupNCCL::WorkNCCL::finishedGPUExecutionInternal() const {
   // hang if another thread is holding the CUDA global context lock. For
   // example, when doing a `cudaDeviceSynchronize` or even
   // `cudaStreamSynchronize`.
+  auto mode = cudaStreamCaptureModeThreadLocal;
+  // push
+  cudaThreadExchangeStreamCaptureMode(&mode);
   if (!ncclEndEvent_->query()) {
+    // pop
+    cudaThreadExchangeStreamCaptureMode(&mode);
     return false;
   }
+  // pop
+  cudaThreadExchangeStreamCaptureMode(&mode);
```

Comment:
nit: shall we create a `watchdogEventQuery` helper (you name it) to hold duplicate code?

Reply:
I think we can avoid this now by just changing the capture mode in the watchdog loop itself---no need to really change the behavior across all `finishedGPUExecutionInternal`/`startedGPUExecutionInternal` calls

Reply:
Ah, that'd be nice! Thank you!